src/tools/rbd_mirror/image_map/Policy.h

```diff
@@ -0,0 +1,185 @@
+    StateTransition::SubActionType get_current_sub_action() const;
+
+    void reset_action();
+
+    void execute_state_callback();
+    void execute_completion_callback(int r);
+
+    inline bool operator==(const Action &rhs) const {
+      return action_type == rhs.get_action_type() &&
+        current_sub_action == rhs.get_current_sub_action();
+    }
+    inline bool operator!=(const Action &rhs) const {
+      return !(*this == rhs);
+    }
+
+  private:
+    Action(StateTransition::ActionType action_type);
+
+    StateTransition::ActionType action_type;                         // action type for this action
+    StateTransition::SubActionType current_sub_action;               // currently executing subaction
+    std::map<StateTransition::SubActionType, Context *> context_map; // map sub action type to context
+                                                                     // callback
+  };
+
+  // init -- called during initialization
+  void init(const std::map<std::string, cls::rbd::MirrorImageMap> &image_mapping);
+
+  // lookup an image from the map
+  std::string lookup(const std::string &global_image_id);
+
+  // shuffle images when instances are added/removed
+  void add_instances(const std::vector<std::string> &instance_ids,
+                     std::set<std::string> *remap_global_image_ids, bool shuffle);
+  void remove_instances(const std::vector<std::string> &instance_ids,
+                        std::set<std::string> *remap_global_image_ids);
+
+  // state machine related methods
+  void init_state(const std::string &global_image_id);
+  void release_state(const std::string &global_image_id);
+
+  bool actions_pending(const std::string &global_image_id);
+
+  bool append_action(const std::string &global_image_id, const Action &action);
+  void prepend_action(const std::string &global_image_id, const Action &action);
+
+  bool start_action(const std::string &global_image_id);
+  bool finish_action(const std::string &global_image_id);
+
+  void sub_action_start(const std::string &global_image_id);
+  bool sub_action_finish(const std::string &global_image_id);
+
+  static const std::string UNMAPPED_INSTANCE_ID;
+
+private:
+  typedef std::list<Action> Actions;
+
+  struct ActionState {
+    Actions actions;                                                  // list of pending actions
+    std::pair<StateTransition::State, StateTransition::State> states; // current,next state
+
+    ActionState()
+      : states(std::make_pair(StateTransition::STATE_UNASSIGNED,
+                              StateTransition::STATE_UNASSIGNED)) {
+    }
+  };
+
+  bool is_shuffle_action(const Action &action) {
+    assert(m_map_lock.is_locked());
+    return action.get_action_type() == StateTransition::ACTION_TYPE_SHUFFLE;
+  }
+
+  librados::IoCtx &m_ioctx;
+  std::map<std::string, ActionState> m_actions;
+  std::map<std::string, utime_t> m_shuffled_timestamp;
+
+  std::set<std::string> m_dead_instances;
```

Comment:
I had to retain this as policy keeps the instance around till all images mapping to an instance are unmapped (migrated) when an instance is removed.